### PR TITLE
composition function types

### PR DESCRIPTION
**Pull request by nullable-eth**
### Description of your changes
Adds meta type for composition `Function`

Fixes # 

I have:

- [X] Read and followed Crossplane's [contribution process].
- [X] Added or updated unit **and** E2E tests for my change.
- [X] Run `make reviewable` to ensure this PR is ready for review.
- [X] Added `backport release-x.y` labels to auto-backport this PR if necessary.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/pkg/meta/v1alpha1/conversion.go [hunk 1]
```diff
@@ -113,3 +118,32 @@ func (p *Provider) ConvertFrom(hub conversion.Hub) error {
...
+	in, ok := hub.(*Function)
+	if !ok {
+		return errors.New(errWrongConvertFromFunction)
+	}
+
+	conv := &GeneratedFromHubConverter{}
+	*f = *conv.Function(in)
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (nullable-eth) [reply apis/pkg/meta/v1alpha1/conversion.go:149]: This and all similar Hub code was mainly added to support [this unit test](https://github.com/crossplane/crossplane/blob/d3a7e39e9264a23c29c3a7dd7f9ae6ba097937cf/internal/xpkg/lint_test.go#L235).  I just kept the function test consistent with the others and that syntax required the hub conversion.

File: internal/xpkg/lint.go [hunk 2]
```diff
@@ -88,6 +90,15 @@ func IsConfiguration(o runtime.Object) error {
...
 }
 
+// IsFunction checks that an object is a Function meta type.
+func IsFunction(o runtime.Object) error {
+	po, _ := TryConvert(o, &pkgmetav1alpha1.Function{})
+	if _, ok := po.(*pkgmetav1alpha1.Function); !ok {
+		return errors.New(errNotMetaFunction)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (nullable-eth) [reply internal/xpkg/lint.go:100]: Done

File: apis/pkg/meta/v1alpha1/function_types.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2020 The Crossplane Authors.
```
[the remark this answers is not in the document]
Comment (nullable-eth) [reply apis/pkg/meta/v1alpha1/function_types.go:2]: Done

File: apis/pkg/meta/v1alpha1/conversion.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (negz) [reply apis/pkg/meta/v1alpha1/conversion.go:149]: Got it. I think we should change how `IsFunction` is implemented and remove this conversion code. Currently it's generating code to convert from `*v1alpha1.Function` to `*v1alpha1.Function` (i.e. the same type) which is a little odd. The other implementations are converting from the `v1alpha1` types to the equivalent `v1` types.

I'll add a comment down on `IsFunction` with a suggestion

File: internal/xpkg/lint.go [hunk 4]
```diff
@@ -88,6 +96,15 @@ func IsConfiguration(o runtime.Object) error {
 	return nil
 }
 
+// IsFunction checks that an object is a Function meta type.
+func IsFunction(o runtime.Object) error {
+	po, _ := TryConvert(o, &pkgmetav1alpha1.Function{})
+	if _, ok := po.(*pkgmetav1alpha1.Function); !ok {
```
Comment (negz) [inline internal/xpkg/lint.go:102]:
```suggestion
	if _, ok := o.(*pkgmetav1alpha1.Function); !ok {
```

I think we can skip conversion here and just test whether the supplied object is a `*pkgmetav1alpha1.Function`. `IsPackage` and `IsConfiguration` need to handle the fact that multiple versions of the same type (i.e. `v1` and `v1alpha1`) exist, but that's not yet the case here. I believe this should allow us to get rid of the conversion code until it's needed.

Comment (nullable-eth) [reply internal/xpkg/lint.go:102]: Yup, that's what I was thinking.

File: apis/pkg/meta/v1alpha1/conversion.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nullable-eth) [reply apis/pkg/meta/v1alpha1/conversion.go]: Removed

**Review by negz — APPROVED** [review]
This looks great, thanks @nullable-eth!

**Pull request merged**